design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,385 @@
+## Background
+
+The Cyberark Secrets Provider for K8s enables customers to use secrets stored and managed in Conjur/DAP vault and consume them as K8s Secrets for their application containers. During Phase 1, an *init* container was deployed in the same Pod as the customer's applications and populate the K8s secrets defined in the customer's manifest. 
+
+### Motivation
+
+The current implementation makes the deployment of the Secrets Provider does not support secret rotation and makes the deployment of the Secrets Provider deeply coupled to the customer's applications which not all customers want.
+
+### Requirement
+
+The solution should support the following capabilities:
+
+1. Deployment, deploy as separate Conjur entity; serve K8s Secrets to many Applications
+2. Support removal of deployment
+
+## Solution
+This solution will be divided into milestones.
+
+- Milestone 1: Serve multiple applications, deploy as as separate entity
+- Milestone 2: Handle rotation, updating K8s Secrets with Conjur secret value
+- Milestone 3: Support reading secrets from files
+- Milestone 4: Listen for K8s Secret creation
+- Milestone 5: Trigger K8s Secret update on Conjur secret value update
+
+For the purposes of this document, we will *only* be addressing the first Milestone.
+
+
+
+### Milestone 1: Serve Multiple Applications
+
+#### Design
+
+The Secrets Provider for K8s will no longer run as an *init* container attached to the application pod, rather run as a separate single Job in a namespace. A Job creates a Pod to perform a task and terminates upon completion. Once the Pod is spun up, it will authenticate to Conjur/DAP in the same way it does today, via authn-k8s. The Pod will then fetch the K8s secret denoted with a label and update them with the Conjur secrets they require.
+
+|      | Deployment                  | Elaboration                                                  | Pros                                                         | Cons                                                         |
+| ---- | --------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| 1    | Run as Job                  | Run as a K8s Job that deploys a Pod that terminates after task completion | - Minimal resources usage<br />- Minimal code changes<br />- Explicit user experience (user expects Job to run to completion quickly) | - "Job" is not an existing application identity granularity. <br /><br />- Deployment will need to be changed (from Job → Deployment) for Milestone 2 (support secrets rotation)<br /><br />- For **deployment**, **deploymentConfig** and **statefulSet** segregation of duties, this solution is not interchangeable with our other secrets retrieval solutions without policy changes. (e.g. secrets provider as init-container, conjur authn k8s secrets, any of our clients or Conjur REST API) |
```

Comment:
> Deployment will need to be changed (from Job → Deployment) for Milestone 2 
I think that in milestone 2 we can say "If you wish to continuously rotate the secrets you should use a Deployment. However, if you prefer a one-time run of the secrets-provider you should use a Job". So it's not really that we change the solution, rather than adding another ability.

Reply:
> For **deployment**, **deploymentConfig** and **statefulSet** segregation of duties, this solution is not interchangeable with our other secrets retrieval solutions without policy changes

I still don't like this sentence and think it is misleading. It is not clear that you are talking about a case where a customer already has a secrets-provider running and they already loaded its corresponding Conjur host. I really think it is a non-issue and can be removed.